subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/OutputChangeDetector.java

```diff
@@ -0,0 +1,144 @@
+
+public class OutputChangeDetector {
+    private final CompareStrategy.ChangeDetector<FileSystemLocationFingerprint> delegate;
+
+    public OutputChangeDetector(CompareStrategy.ChangeDetector<FileSystemLocationFingerprint> delegate) {
+        this.delegate = delegate;
+    }
+
+    public boolean visitChangesSince(FileSystemSnapshot previous, FileSystemSnapshot current, String propertyTitle, ChangeVisitor visitor) {
+        if (previous == current) {
+            return true;
+        }
+        if (previous == FileSystemSnapshot.EMPTY) {
+            return visitAllChanges(current, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.added(
+                snapshot.getAbsolutePath(),
+                propertyTitle,
+                createFingerprint(relativePath, snapshot)
+            ));
+        } else if (current == FileSystemSnapshot.EMPTY) {
+            return visitAllChanges(previous, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.removed(
+                snapshot.getAbsolutePath(),
+                propertyTitle,
+                createFingerprint(relativePath, snapshot)
+            ));
+        } else {
+            FileSystemLocationSnapshot previousSnapshot = (FileSystemLocationSnapshot) previous;
+            FileSystemLocationSnapshot currentSnapshot = (FileSystemLocationSnapshot) current;
+            if (previousSnapshot.getHash().equals(currentSnapshot.getHash())) {
+                return true;
+            } else if (currentSnapshot.getType() != FileType.Directory || previousSnapshot.getType() != FileType.Directory) {
+                // The hash is different and one of the two snapshots is not a directory falls into cases:
+                // 1. The type of the snapshot is different.
+                // 2. The content hash of a regular file changed.
+                // In any case, the root needs to be reported as changed:
+                return visitor.visitChange(FINGERPRINT_CHANGE_FACTORY.modified(
+                    currentSnapshot.getAbsolutePath(),
+                    propertyTitle,
+                    createRootFingerprint(previousSnapshot),
+                    createRootFingerprint(currentSnapshot)
+                )) &&
+                    // Only one of the types will be a directory, so we now need to report all children of the other one.
+                    visitAllChildChanges(previousSnapshot, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.removed(
+                        snapshot.getAbsolutePath(),
+                        propertyTitle,
+                        createFingerprint(relativePath, snapshot)
+                    )) &&
+                    visitAllChildChanges(currentSnapshot, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.added(
+                        snapshot.getAbsolutePath(),
+                        propertyTitle,
+                        createFingerprint(relativePath, snapshot)
+                    ));
```

Comment:
Can we break this up into separate `return`s maybe? I wonder if we could improve the back-and-forth between `boolean` and `SnapshotVisitResult` as well a bit that way.

Reply:
I broke this up into separate returns. It doesn't seem much better. Did you have something else in mind?

Reply:
I tried it also, but didn't manage to make it better either. Let's leave as is.